File: brickadia-ue4ss-re/scripts/ghidra_dump_anchor_xrefs.py
# Dump xrefs for known CL12960 UTF-16 anchor addresses without scanning the whole program.
# @category Brickadia

ANCHORS = [
    ("StaticFindObjectFast", "0x145d372c0"),
    ("FUObjectHashTables", "0x145d3eeae"),
    ("HashOuter", "0x145d3ed64"),
]


def root_function(address):
    fn = getFunctionContaining(address)
    if fn is None:
        return None
    return fn


def describe_function(fn):
    if fn is None:
        return "no function"
    return "{} @ {}".format(fn.getName(True), fn.getEntryPoint())


def print_anchor(name, address_text):
    addr = toAddr(address_text)
    print("== Anchor: {} @ {} ==".format(name, addr))
    refs = getReferencesTo(addr)
    seen = set()
    count = 0
    for ref in refs:
        from_addr = ref.getFromAddress()
        fn = root_function(from_addr)
        key = (str(from_addr), fn.getEntryPoint().toString() if fn else None)
        if key in seen:
            continue
        seen.add(key)
        count += 1
        print("  ref {} -> {}".format(from_addr, describe_function(fn)))
    if count == 0:
        print("  no references")
    print("")


print("Program: {}".format(currentProgram.getName()))
print("")
for name, address_text in ANCHORS:
    print_anchor(name, address_text)

File: brickadia-ue4ss-re/scripts/GhidraDumpAnchorXrefs.java
// Dump xrefs for known CL12960 UTF-16 anchor addresses without scanning the whole program.
// @category Brickadia

import ghidra.app.script.GhidraScript;
import ghidra.program.model.address.Address;
import ghidra.program.model.listing.Function;
import ghidra.program.model.symbol.Reference;

public class GhidraDumpAnchorXrefs extends GhidraScript {

	private static final String[][] ANCHORS = new String[][] {
		{ "StaticFindObjectFast", "0x145d372c0" },
		{ "FUObjectHashTables", "0x145d3eeae" },
		{ "HashOuter", "0x145d3ed64" }
	};

	private String describeFunction(Function function) {
		if (function == null) {
			return "no function";
		}
		return function.getName(true) + " @ " + function.getEntryPoint();
	}

	private void printAnchor(String name, String addressText) throws Exception {
		Address address = toAddr(addressText);
		printf("== Anchor: %s @ %s ==\n", name, address);
		Reference[] refs = getReferencesTo(address);
		if (refs.length == 0) {
			printf("  no references\n\n");
			return;
		}

		java.util.HashSet<String> seen = new java.util.HashSet<>();
		for (Reference ref : refs) {
			Address fromAddress = ref.getFromAddress();
			Function function = getFunctionContaining(fromAddress);
			String key = fromAddress.toString() + "|" + (function == null ? "" : function.getEntryPoint().toString());
			if (!seen.add(key)) {
				continue;
			}
			printf("  ref %s -> %s\n", fromAddress, describeFunction(function));
		}
		printf("\n");
	}

	@Override
	protected void run() throws Exception {
		printf("Program: %s\n\n", currentProgram.getName());
		for (String[] anchor : ANCHORS) {
			printAnchor(anchor[0], anchor[1]);
		}
	}
}
